src/blocks/scratch3_event.js

```diff
@@ -12,23 +12,76 @@ function Scratch3EventBlocks(runtime) {
  */
 Scratch3EventBlocks.prototype.getPrimitives = function() {
     return {
-        'event_whenflagclicked': this.whenFlagClicked,
-        'event_whenbroadcastreceived': this.whenBroadcastReceived,
-        'event_broadcast': this.broadcast
+        'event_broadcast': this.broadcast,
+        'event_broadcastandwait': this.broadcastAndWait,
+        'event_whengreaterthan': this.hatGreaterThanPredicate
     };
 };
 
+Scratch3EventBlocks.prototype.getHats = function () {
+    return {
+        'event_whenflagclicked': {
+            restartExistingThreads: true
+        },
+        /*'event_whenkeypressed': {
+            restartExistingThreads: false
+        },
+        'event_whenthisspriteclicked': {
+            restartExistingThreads: true
+        },
+        'event_whenbackdropswitchesto': {
+            restartExistingThreads: true
+        },*/
+        'event_whengreaterthan': {
+            restartExistingThreads: false,
+            edgeTriggered: true
+        },
+        'event_whenbroadcastreceived': {
+            restartExistingThreads: true
+        }
+    };
+};
 
-Scratch3EventBlocks.prototype.whenFlagClicked = function() {
-    // No-op
+Scratch3EventBlocks.prototype.hatGreaterThanPredicate = function (args, util) {
+    // @todo: Other cases :)
+    if (args.WHENGREATERTHANMENU == 'TIMER') {
+        return util.ioQuery('clock', 'projectTimer') > args.VALUE;
+    }
+    return false;
 };
 
-Scratch3EventBlocks.prototype.whenBroadcastReceived = function() {
-    // No-op
+Scratch3EventBlocks.prototype.broadcast = function(args, util) {
+    util.triggerHats('event_whenbroadcastreceived', {
+        'BROADCAST_OPTION': args.BROADCAST_OPTION
+    });
 };
 
-Scratch3EventBlocks.prototype.broadcast = function() {
-    // @todo
+Scratch3EventBlocks.prototype.broadcastAndWait = function (args, util) {
+    // Have we run before, triggering threads?
+    if (!util.stackFrame.triggeredThreads) {
+        // No - trigger hats for this broadcast.
+        util.stackFrame.triggeredThreads = util.triggerHats(
+            'event_whenbroadcastreceived', {
+                'BROADCAST_OPTION': args.BROADCAST_OPTION
+            }
+        );
+        if (util.stackFrame.triggeredThreads.length == 0) {
+            // Nothing was started.
+            return;
+        }
+    }
+    // We've run before; check if the wait is still going on.
+    var waiting = false;
+    for (var i = 0; i < util.stackFrame.triggeredThreads.length; i++) {
+        var thread = util.stackFrame.triggeredThreads[i];
+        var activeThreads = this.runtime.threads;
+        if (activeThreads.indexOf(thread) > -1) { // @todo: A cleaner way?
+            waiting = true;
+        }
+    }
```

Comment:
Sure, this is cleaner:

```js
for (let thread of util.stackFrame.triggeredThreads) {
  if (this.runtime.threads.includes(thread)) {
    waiting = true;
  }
}
```

"For each of the triggered threads, check if that thread is included in the running threads, and if so set waiting to true." Makes sense to me.

It's really the exact same thing as the current code, but easier to read.. (But it would require a new polyfill - probably not worth it, since obviously you can just use `..indexOf(..) >= 0`!)